src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -168,6 +219,221 @@ private static IDataTransform Create(IHostEnvironment env, Arguments args, IData
+                if (type.ItemType != expectedType)
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", inputsForTraining[i], expectedType.ToString(), type.ToString());
+            }
+
+            var fetchList = new List<string>();
+            if (args.LossOperation != null)
+                fetchList.Add(args.LossOperation);
+            if (args.MetricOperation != null)
+                fetchList.Add(args.MetricOperation);
+
+            for (int epoch = 0; epoch < args.Epoch; epoch++)
+            {
+                using (var cursor = input.GetRowCursor(a => true))
+                {
+                    var srcTensorGetters = GetTensorValueGetters(cursor, inputColIndices, isInputVector, tfInputTypes, tfInputShapes);
+
+                    float loss=0;
+                    float metric=0;
+                    int rows = 0;
+                    bool isDataLeft = false;
+                    using (var ch = _host.Start("Training TensorFlow model..."))
+                    {
+                        while (cursor.MoveNext())
+                        {
+                            rows++;
+                            for (int i = 0; i < inputColIndices.Length; i++)
+                            {
+                                isDataLeft = true;
+                                srcTensorGetters[i].BufferTrainingData();
+                            }
+
+                            if ((rows % args.BatchSize) == 0)
+                            {
+                                isDataLeft = false;
+                                (loss, metric) = TrainBatch(inputColIndices, inputsForTraining, srcTensorGetters, fetchList, args);
+                            }
+                        }
+                        if (isDataLeft)
+                        {
+                            isDataLeft = false;
+                            ch.Warning("Not training on the last batch. The batch size is less than {0}.", args.BatchSize);
+                        }
+                        ch.Info("Loss: {0}, Metric: {1}", fetchList.Count > 0 ? loss.ToString("#.####") : "None",
+                                fetchList.Count > 1 ? metric.ToString("#.####") : "None");
+                    }
+                }
+            }
+            UpdateModelOnDisk(model);
+        }
+
+        /// <summary>
+        /// Updates the model on the disk.
+        /// After retraining Session and Graphs are both up-to-date
+        /// However model on disk is not which is used to serialzed to ML.Net stream
+        /// </summary>
+        private void UpdateModelOnDisk(string modelDir)
+        {
+            // Save the model on disk
+            var path = Path.Combine(modelDir, "mlnet_model");
+            Session.GetRunner().AddInput("save/Const", TFTensor.CreateString(Encoding.UTF8.GetBytes(path)))
+                    .AddTarget("save/control_dependency").Run();
```

Comment:
Few questions:

1. is this what's serializing the Session to disk ?  

2. it will save only the parameter weights, not the saved_model.pb graph itself .. am i right ?

3. Is this a standard technique -- of using nodes with names  "save/Const" and "save/control_dependency"  to overwrite model weights

Reply:
Yes to all three question above.
TensorFlow saves/restores models through these operations on graph.

---
In reply to: [221098961](https://github.com/dotnet/machinelearning/pull/1063#discussion_r221098961) [](ancestors = 221098961)

Reply:
for #3, does TF document this somewhere in code / other language bindings etc.  

Just using a combination of strings  "save/Const" and and "save/control_dependency"  seems so arbitrary  :)  .. 

wonder if this is the same technique used when checkpointing models ?

---
In reply to: [221102898](https://github.com/dotnet/machinelearning/pull/1063#discussion_r221102898) [](ancestors = 221102898,221098961)

Reply:
Yes, this is how TensorFlow  saves/restores models to/from disks. Both for checkpointing and simple_save. There is a reason to inject those operation in the graph because TF prefers all operation through graph instead of doing it through python or other language code. I will find a proper documentation for that. 

I have made a parameter in `TrainingArgument` class so that in case user need to change it they can change. However, it is highly unlikely to get changed. 

---
In reply to: [221104033](https://github.com/dotnet/machinelearning/pull/1063#discussion_r221104033) [](ancestors = 221104033,221102898,221098961)